### DISCUSSION TITLE
Cliente. Procesar en bloque vs cuadrícula individual

### DISCUSSION THREAD

**Discussion by jarufe** [manudona92/MapantES#70] [Ideas]
Evolución de los datos LiDAR implica mayor volumen de datos. Un mejor rendimiento pasa por un mejor procesamiento de los datos, pero también por otros aspectos como la descarga del material a procesar.

### Ejemplos de tamaños de ficheros LAZ segunda cobertura
  Galicia W. 2015. 0,5 ptos/m2. 30 MB
  Galicia E. 2016. 0,5 ptos/m2. 50 MB
  Aragón N. 2021. 1 pto/m2. 90 MB
  Extremadura N. 2018-2019. 2 ptos/m2. 170 MB
  Navarra. 2017. 14 ptos/m2. 300 MB

Ya en segunda cobertura se están observando zonas que tienen tamaños de ficheros mayores que antes. Fácilmente nos encontramos con LAZ de entre 150 y 200 MB.

### Idea para el replanteamiento del procesamiento por parte de cada cliente
  Asignación y procesamiento por bloques de cuadrículas en lugar de cuadrículas individuales
    
![image](https://github.com/manudona92/MapantES/assets/17064412/f52eba3f-8b77-44bc-ab5f-6dc391bacbc3)

  El funcionamiento podría ser mixto: clientes procesando cuadrículas individuales, clientes por bloque
  La definición de bloques (y qué cuadrículas le pertenecen) podrían estar preparados de antemano. Tal vez se pueda establecer un geoproceso relativamente sencillo que permita dar de alta registros de bloques y sus cuadrículas
  A un cliente se le asigna un bloque y procesaría todas las cuadrículas de ese bloque, teniendo que descargar los ficheros afectados para todo el bloque (al comienzo o a medida que vaya necesitando ficheros)

### Ejemplos de bloques y sus cuadrículas
**Procesado de 9 (3x3) cuadrículas**
  Por cada cuadrícula se necesitan hasta 4 ficheros
Si tomadas de forma individual
  Descargar 4 ficheros cada vez que se procese una cuadrícula => 9x4 = 36 ficheros
Si asignadas en bloque
  Descargar 4x4 = 16 ficheros y procesar las 9 cuadrículas

**Si 36 (6x6) cuadrículas**
Individual
  => 36x4 = 144 ficheros (en general, si n x n cuadrículas, necesarios hasta n x n x 4 ficheros)
Bloque
  => 7x7 = 49 ficheros (en general, si n x n cuadrículas, necesarios hasta (n+1) x (n+1) ficheros)

**En cuanto al volumen de datos para descargar (solo la descarga, no procesamiento)**
Si 30MB por fichero
  Individual => 4320 MB
  Bloque => 1470 MB
  Si 6" por fichero, individual => 14', bloque => 5'
Si 100MB por fichero
  Individual => 14400 MB
  Bloque => 4900 MB
  Si 20" por fichero, individual => 48', bloque => 16'

Si toda España fueran 14000 bloques con ficheros de 100MB
  Individual => 11200h = 466días, bloque => 3733h = 155días
Si 50 instancias funcionando
  Individual => 9días, bloque => 3días

### Un poco más de detalle en el posible flujo

- A un cliente se le asigna un bloque, con lo que quedarían preasignadas todas las cuadrículas del bloque, no pudiendo ser asignadas a otros clientes. En realidad, buscando la sencillez se podría evitar esta restricción. Si a un cliente de procesamiento individual se le asigna una cuadrícula de un bloque, cuando llegue el momento de que quien procesa el bloque tenga que procesar esa cuadrícula, consultando el estado previamente, el cliente ya vería que está finalizada y pasaría de ella
- El bloque pasaría a estado asignado
- Cada vez que se terminara el procesamiento de una cuadrícula, esa cuadrícula se marcaría como finalizada
- Al finalizar todas las cuadrículas del bloque, el bloque se marcaría como finalizado
- Borrar todos los ficheros del bloque en local
- Cuanto más grande sea el bloque
  - Menos ficheros a descargar
  - Más riesgo de que si se corta el procesamiento, queden más cuadrículas en estado asignado pero que no se van a procesar
- Si durante un bloque se corta su procesamiento de forma abrupta
  - Las cuadrículas ya procesadas quedaron marcadas como finalizadas
  - El bloque, pasado un tiempo, podría marcarse como asignable nuevamente
  - Si un cliente recibe ese bloque, antes de procesar una cuadrícula se comprobaría si ya está finalizada, continuando con la siguiente cuadrícula en caso afirmativo
- Procesando un bloque
  - No habría por qué descargar todos los ficheros de antemano
  - Para una cuadrícula
    - Consultar qué ficheros son necesarios
    - Descargar un fichero si todavía no existe en local
- En el programa cliente se podría poner una casilla para que el usuario marque si está de acuerdo en hacer un procesamiento de larga duración, es decir, que marque si quiere un procesamiento de bloques

**Comment by agustincaballero** [comment]
Hola, Javi. Muy interesante las ideas que expones aquí.

Como bien dices, ahora mismo el cliente procesa cuadrículas de 1x1km, mientras que los ficheros LAZ son de 2x2km; cuando un cliente procesa una cuadrícula se tiene que descargar entre 1 y 4 ficheros LAZ, según sea el solapamiento. Por tanto, si procesa una y a continuación procesa otra contigua, se puede estar descargando varias veces el mismo fichero LAZ.

La forma más eficiente de abordar esto sería, como también comentas, que el cliente se bajara y procesara varias de golpe (es lo que has llamado cliente por bloques); pero esto supone modificar bastante el "workflow" que tenemos actualmente, y tiene también algunas desventajas; pero es estudiable a largo plazo (tal vez para la tercera cobertura).

Una forma más sencilla y factible a corto/medio plazo es la que también comentas: que a un cliente se le asignen preferentemente cuadrículas contiguas a otras que acaba de procesar; antes de descargar el LAZ, comprobaría si ya lo tiene en el disco duro y ahorraría esta descarga. Esto implica que los LAZ no se borrarían inmedatamente; habría que establecer un tiempo de borrado, o tal vez un tamaño máximo que el usuario quiere que ocupe este almacén temporal. Para que sea efectivo, tendría que ser bastante grande (¿5 o 10 GB?), así que habrá que efectivamente informar o pedir permiso al usuario.

Esta modificación del cliente la veo bastante factible, voy a estudiarla detenidamente.

Os iré informando de los avances.

**Reply by jaluzdi** [reply]
Buenas noches,

Llevo dándole vueltas a la estupenda idea del procesamiento por bloques, que puso Javi hace una semana, ya que ahora con el rusty-pullauta, por ejemplo actualmente al procesar Asturias, si se tiene que bajar cuatro cuadriculas del IGN (de 2km x 2km, como comenta Agustín), para acabar generando una única cuadrícula de 1km x 1km, al final casi la mitad del tiempo de proceso total, se va en esas descargas y generación de la cuadrícula final a procesar por el cliente mapant.

Programar ese tipo de procesamiento por bloques, como comenta Agustín , imagino que además llevará aparejado muchos cambios en el workflow actual, por todo el tema de la gestión de los bloques, interrupciones repentinas en el procesamiento de un bloque, ....

Por aportar una idea diferente, y seguramente disparatada ... aunque ya me imagino que igual se la propondriais al IGN en su momento, al inicio del proyecto.

La solución más eficiente, para nosotros sería que el propio IGN generase esas cuadrículas de 1km x 1km, al tener los datos en sus propios servidores, con lo cual el tiempo de descarga de dichos archivos *.laz sería casi nulo para ellos.

Imagino que el IGN tendrá servidores LINUX. o UNIX.

Dentro de las LAStools hay algunas herramientas que son de código libre. Es decir, que aunque oficialmente solo se distribuyan ejecutables para Windows de todas ellas, hay unas pocas herramientas de las cuales su código fuente está disponible para ser compilado en cualquier sistema operativo.

Y justo una de las herramientas libres es **las2las**, que creo que es la que se usa para generar las cuadrículas de 1km x 1km.
Sólo tendrían que compilar el código fuente de las2las en su servidor y generar el ejecutable, y ya podrían generar las cuadrículas 1x1.

A todo este planteamiento, seguramente dirán que para que quieren volver a generar esas cuadrículas así (1x1) cuando ya las tienen en otro tamaño (2x2) con el consiguiente gasto de espacio en disco duro en sus servidores. Que si alguien las quiere en otro tamaño, que se les procese. Ahí radicaría la mayor pega que podrían poner. Y bastante entendible, por otro lado.

Se les podrían plantear soluciones intermedias. Es decir, por ejemplo, en vez de tener toda España procesada y convertida en cuadrículas 1x1, que sólo fueran generando aquellas que se fuesen a procesar por los clientes de mapant (ya que sólemos procesar por lotes). Y una vez procesadas por mapant, que las borre de su servidor. Por ejemplo, si la península son 500.000 km2, que fueran generando lotes de 10.000 - 20.000 km2

De esta forma, casi no habría que modificar el workflow actual. En todo caso lo simplificaría, ya que en vez de descargar de 1 a 4 ficheros laz de 2x2 para generar una cuadricula 1x1, ... con el nuevo proceso sólo habría que descargar la cuadrícula a procesar y listo.

Y otra opción, si el IGN se negase en redondo (cosa bastante comprensible por otro lado), podría ser que lo que planteo, en vez de hacerlo el IGN, lo hiciera un servidor Linux dedicado solo a ello. Descargar cuadriculas del IGN y formar las cuadrículas 1x1 a procesar. Que en un momento dado igual hasta lo podría llegar a hacer el propio servidor donde está alojado mapant, por ejemplo, nocturnamente que imagino que de madrugada, atenderá pocas peticiones y estará relativamente ocioso :smile:

**Reply by agustincaballero** [reply]
Hola. Son ideas interesantes, pero dudo muchísimo que el IGN pueda hacer eso. Una provincia media (10000km2) ocupará al menos 50Mbx10000=500Gb; España entera serían 25Tb.
Además hay que tener en cuenta que no basta con sacar LAZ de 1x1 km, porque hay que añadirle 50 metros a cada lado de buffer para que Karttapullautin haga bien los bordes. Y por tanto no basta con trocear un laz de 2x2 en trozos de 1x1, sino que hay que procesar también las celdas contiguas.
Por otra parte, el hacerlo "a trozos" supone que alguien (nosotros o ellos) tenga que repetir 50 veces el proceso, lo veo complicado y que requiere de mucho trabajo.

**Comment by agustincaballero** [comment]
Buenas, dije que os informaría del proceso de modificaciones, en el cliente, y aquí van:

Lo que tengo hecho consiste en que al descargar los laz, éstos se van a almacenar en el disco duro, y en sucesivos procesamientos se usarán esos laz en lugar de descargarlos. 

La "pega" es que el cliente tiene que reservar un espacio considerable (¿2 Gb?) extra para almacenar esta caché.

No sé cómo irá la gente en general de disco duro, ¿cómo lo veis?

**Reply by jaluzdi** [reply]
Buenos días,

Personalmente lo veo bien. Avisando a los/as voluntarios/as de que se requiere ese espacio para hacer el proceso más eficiente.
Estaría bien explicar el porque de esa reserva de espacio. Con un enlace a la explicación de Javi en el GitHub bastaría. Que está muy bien explicado. 
En los tiempos que corren, 2 GB de disco duro no es que sea gran cosa.
Incluso para quien lo estuviera corriendo en maquinas "cloud" tipo AWS o Ms Azure, creo que el disco duro mínimo que te ponen son 30 GB.
Diría que no debiera haber problemas en los PCs de los/as voluntarios/as (en mi caso ninguno)
